draft-wang-ppm-differential-privacy.md

```diff
@@ -403,31 +403,266 @@ This section describes various mechanisms required for implementing DP
+However, in case the minimum batch size is not met when Aggregators want to
+output their aggregate shares, each Aggregator can sample the same randomized
+response Client-DP on the "missing" Clients with zero-hot vectors, so the
+central DP guarantee can still be satisfied, based on the privacy amplification
+analysis in {{FMT20}} and {{FMT22}}.
+
+~~~
+class Prio3SumVecWithRandomizedResponse:
+    Measurement = Vec[Unsigned]
+    AggregateShare = Vec[Field]
+    AggregateResult = Vec[Unsigned]
+    DebiasedAggregateResult = Vec[float]
+
+    def __init__(self,
+                 rr_mechanism: DpMechanism):
+        self.rr_mechanism = rr_mechanism
+
+    def add_noise_to_measurement(self,
+                                 meas: Measurement,
+                                 ) -> Measurement:
+        """
+        Apply randomized response mechanism to measurement.
+        """
+        return self.rr_mechanism.add_noise(meas)
+
+    def add_noise_to_agg_share(self,
+                               agg_share: AggregateShare,
+                               meas_count: Unsigned,
+                               min_batch_size: Unsigned,
+                               ) -> AggregateShare:
+        """
+        If minimum batch size is met, this function does nothing,
+        because DP guarantee would have already been met.
+        Otherwise, add the same randomized response noise on the
+        missing Clients.
+        """
+        if meas_count >= batch_size:
+            return agg_share
+        noise = self.rr_mechanism.sample_noise(
+            batch_size - meas_count, len(agg_share)
+        )
+        return [a + Field(b) for (a, b) in zip(agg_share, noise)]
+
+    def debias_agg_result(self,
+                          agg_result: AggregateResult,
+                          meas_count: Unsigned,
+                          min_batch_size: Unsigned,
+                          num_aggregators: Unsigned,
+                          ) -> DebiasedAggregateResult:
+        """
+        Debias aggregate result with randomized response.
+        """
+        if meas_count >= min_batch_size:
+            return self.rr_mechanism.debias(agg_result, meas_count)
+        # The count passed to debiasing will be:
+        # the number of actual Clients + the "missing" Clients
+        # sampled by all Aggregators.
+        debiasing_count = \
+            meas_count + \
+            num_aggregators * (min_batch_size - meas_count)
+        return self.rr_mechanism.debias(agg_result, debiasing_count)
+~~~
+
+> TODO(issue #10): replace `rr_mechanism` once we have concretely defined
+> it in {{rr}}.
+
+### Prio3Histogram with Discrete Gaussian
+
+For Prio3Histogram, we will apply an Aggregator-only DP mechanism,
+that is implemented with discrete Gaussian.
+
+~~~
+class Prio3HistogramWithDiscreteGaussian:
+    Measurement = Unsigned
+    AggregateShare = Vec[Field]
+    AggregateResult = Vec[int]
```

Comment:
There is some small caveat about the aggregate result type for `Prio3Histogram` with discrete Gaussian, because discrete Gaussian can sample negative noise, so it's possible that aggregate result becomes negative. In the Field world, it is at the tail end of field modulus.